File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/Before/src/test/java/com/wimp/specs/drivers/AuthenticationApiDriver.java
package com.wimp.specs.drivers;

import com.wimp.app.restapi.LoginRequest;
import com.wimp.app.restapi.LoginResponse;
import com.wimp.specs.support.LambdaAction;
import com.wimp.specs.support.RestApiContext;
import com.wimp.specs.support.TestAction;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;

@Component
public class AuthenticationApiDriver {
    private final RestApiContext restApiContext;

    public AuthenticationApiDriver(RestApiContext restApiContext) {
        this.restApiContext = restApiContext;
    }

    public TestAction<LoginResponse> login(String customerName, String password) {
        return new LambdaAction<>(() -> {
            LoginResponse response = restApiContext.processRequest(
                "Login",
                HttpMethod.POST,
                "/api/auth/login",
                new LoginRequest(customerName, password),
                HttpStatus.OK,
                LoginResponse.class);
            restApiContext.setBearerToken(response.token());
            return response;
        });
    }
}

File: Reqnroll/Chapter15/15.5-AttemptActionPattern/After/WIMP.Specs/Drivers/AuthenticationApiDriver.cs
using WIMP.App.RestApi;
using WIMP.Specs.Support;

namespace WIMP.Specs.Drivers;

public class AuthenticationApiDriver(RestApiContext restApiContext)
{
    public TestAction<LoginResponse> Login(string customerName, string password) =>
        new LambdaAction<LoginResponse>("Login", async () =>
        {
            var loginResponse = await restApiContext.ProcessRequest<LoginResponse>(
                "Login", HttpMethod.Post, "/api/auth/login",
                new LoginRequest(customerName, password));
            restApiContext.BearerToken = loginResponse.Token;
            return loginResponse;
        });
}
